### PR TITLE
Add Iterator::intersperse

### PR DESCRIPTION
**Pull request by jonas-schievink**
This is occasionally useful when wanting to separate an iterator's items.

Implementation adapted from [itertools](https://github.com/rust-itertools/itertools).

Drawback: Breaks all code currently using itertools' `intersperse` method since the method name collides.

### CODE REVIEW

**Review by JulianKnodt — COMMENTED** [review]
Is there any benefit of `intersperse` over something like:
```
let range = 0..100;
let a = repeat(3);
range.zip(a).flat_map(|(i, j)| {
    once(i).chain(once(j))
});
```
The only difference would be that there would be an extra element at the end in this case.

Edit: Sorry for the late edit, but I realize the fold portion of this could also be represented as
`.fold_first(|acc, next| f(acc, interspersed_element, next))`. You can also create an iterator by taking the first element off and prefixing the rest with the repeated one, which handles the case where you need an interspersed iterator. I think that this iterator can be represented by other more primitive iterators, but if it is a common use case I think it could be useful.

**Comment by scottmcm** [comment]
> Drawback: Breaks all code currently using itertools' intersperse method since the method name collides.

I really [wish](https://internals.rust-lang.org/t/idea-paths-in-method-names/6834?u=scottmcm) we had a nice way to let people disambiguate extension method conflicts 🙁

File: library/core/src/iter/adapters/intersperse.rs
Comment (ollie27) [inline library/core/src/iter/adapters/intersperse.rs:69]:
```suggestion
        let hi = hi.and_then(|hi| hi.checked_add(hi)).and_then(|hi| hi.checked_add(has_peek));
```

If `hi` overflows it should be `None` not `Some(usize::MAX)`.

**Comment by Mark-Simulacrum** [comment]
I'm going to r? @LukasKalbertodt as a libs team member. The implementation here looks good to me.

I personally suspect that it might not be a good idea to land this given the itertools conflict, but I did have an idea for how we could alleviate that in the future at least: itertools could have `#[cfg(not(rust_has_intersperse))]` on its extension trait method, set in build.rs via something like https://docs.rs/autocfg/1.0.1/autocfg/ or ideally cfg accessible (though that would need the feature stabilized and https://github.com/rust-lang/rust/issues/72011 fixed).

That doesn't help us now though.

**Comment by pickfire** [comment]
I think we could itertools could use `#[cfg(path = "Iterator::intersperse")]`.

**Review by jswrenn — COMMENTED** [review]
So, I'm the current maintainer of Itertools. I'm *thrilled* to see this PR! `Intersperse` is widely useful, and very deserving of being a method on `Iterator`. My very first PR to Itertools was actually optimizing `Intersperse::fold`, because `chain`s of `Intersperse`ed tokens are widespread in the internals of the standard proc macro libraries.

> I personally suspect that it might not be a good idea to land this given the itertools conflict

I really wouldn't want Itertools to get in the way of a good addition. As an Itertools _user_, the fixes for these sorts of conflicts are pretty simple: you just used the fully-qualified method call syntax, or limit the scope of your import of `Itertools` so that your call to libcore's `intersperse` is unaffected. (For the uninitiated, the introduction of `Iterator::flatten` caused the same issues. I don't think it posed much more than a minor inconvenience.)

It's only as an Itertools _maintainer_ that these sorts of conflicts cause me a real headache. I really don't like that our README has to *discourage* contributions to minimize these sorts of issues, and I don't like not merging PRs that seem *so* useful that I can foresee them making their way into `Iterator`.

On that note, our upcoming 0.10.0 release is going to include [`intersperse_with`](https://github.com/rust-itertools/itertools/pull/381), a generalized form of `intersperse`. If `Iterator` is going to gain `intersperse`, it doesn't seem out of the question that it might want `intersperse_with` too. Should I cut it from the release to avoid a potential future conflict?

> in the future at least: itertools could have #[cfg(not(rust_has_intersperse))] on its extension trait method, set in build.rs via something like https://docs.rs/autocfg/1.0.1/autocfg/

This seems promising(!!!), but maybe a little unpleasant for contributors. Any new, non-allocating method added to `Itertools` poses a potential future conflict with `Itertools`. We'd need to test for *every* such method name in our `build.rs`. [I'd much prefer a permanent solution to this problem.](https://github.com/rust-lang/rfcs/pull/2845)

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [reply library/core/src/iter/adapters/intersperse.rs:42]: But I see that other specializations do include `try_fold` too. I think there may be a reason they do it.

Comment (the8472) [reply library/core/src/iter/adapters/intersperse.rs:42]: There are several forwarding hierarchies in `Iterator`. `fold` and `try_fold` are separate ones.
See the (now somewhat outdated) analysis in the PR that made `fold` independent from `try_fold`: https://github.com/rust-lang/rust/pull/72139#issuecomment-627612962

Comment (pickfire) [reply library/core/src/iter/adapters/intersperse.rs:42]: @the8472 So it means we should implement all the important ones?

Comment (the8472) [reply library/core/src/iter/adapters/intersperse.rs:42]: If the custom `fold` implementation provides a speedup over the default one which relies on `next` then the other ones would likely benefit too.
Additionally the other optimization and extension traits that most iterator adapters implement such as `ExactSizeIterator`, `TrustedLen`, `DoubleEndedIterator`, `FusedIterator`, `TrustedRandomAccess` all seem applicable here.

File: library/core/src/iter/adapters/intersperse.rs
Comment (the8472) [inline library/core/src/iter/adapters/intersperse.rs:21]: Is that a good idea? Then merely constructing the iterator will advance the wrapped one. `Peekable` which is similar doesn't do this.
It violates [laziness](https://doc.rust-lang.org/std/iter/index.html#laziness) property of iterator adapters.

File: library/core/src/iter/adapters/intersperse.rs
Comment (jonas-schievink) [reply library/core/src/iter/adapters/intersperse.rs:44]: Added an implementation

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [inline library/core/src/iter/adapters/intersperse.rs:65]: Would it be better to / Is it possible?
```suggestion
        self.iter.fold(accum, |accum, x| f(f(accum, element.clone()), x))
```

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [inline library/core/src/iter/adapters/intersperse.rs:98]: Why shouldn't it subtract first and then add? This could allow one more value on the upper bound if it hit `usize::MAX`. Will `lo` and `hi` every be `0`?
```suggestion
        let lo = lo.saturating_sub(next_is_elem as usize).saturating_add(lo);
        let hi = hi
            .and_then(|hi| hi.checked_sub(next_is_elem as usize))
            .and_then(|hi| hi.checked_add(hi));
```

File: library/core/src/iter/traits/iterator.rs
Comment (pickfire) [inline library/core/src/iter/traits/iterator.rs:549]: Add different example and some humor (not tested)
```suggestion
    /// let s = ["996", "icu"].iter().copied().intersperse(".").collect::<String>();
    /// assert_eq!(s, "996.icu");
    ///
    /// let v = [&[9, 9, 6], &[6, 6, 9], &[1, 1, 2]];
    /// let v = v.iter().copied().intersperse(&[0]).collect::<Vec<_>>();
    /// assert_eq!(v, &[9, 9, 6, 0, 6, 6, 9, 0, 1, 1, 2]);
```

File: library/core/src/iter/adapters/intersperse.rs
Comment (ollie27) [inline library/core/src/iter/adapters/intersperse.rs:68]: `try_fold` will need to keep `needs_sep` up to date so if it short circuits the iterator will continue to output the correct items.

File: library/core/src/iter/adapters/intersperse.rs
Comment (ollie27) [reply library/core/src/iter/adapters/intersperse.rs:98]: For `hi`, in addition to doing the subtraction first, it should use `saturating_sub` not `checked_sub` because if `hi` is `Some(0)` it should stay `Some(0)`.

Comment (pickfire) [reply library/core/src/iter/adapters/intersperse.rs:98]: Would it be better to store `next_is_elem as usize` as a variable? Does it affect compilation?

File: library/core/src/iter/adapters/intersperse.rs
Comment (jonas-schievink) [reply library/core/src/iter/adapters/intersperse.rs:68]: Yeah, you're right. That isn't even enough when the `f` call with the separator fails though. I'd somehow have to put the `x` item back into the iterator.

I'll remove `try_fold` for now. Optimizations can be implemented later, and it isn't even clear to me whether this will land or not.

File: library/core/src/iter/adapters/intersperse.rs
Comment (jonas-schievink) [reply library/core/src/iter/adapters/intersperse.rs:65]: I don't think this is more readable

File: library/core/src/iter/adapters/intersperse.rs
Comment (jonas-schievink) [reply library/core/src/iter/adapters/intersperse.rs:98]: Fixed

File: library/core/src/iter/traits/iterator.rs
Comment (jonas-schievink) [reply library/core/src/iter/traits/iterator.rs:549]: I'd rather not

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [inline library/core/src/iter/adapters/intersperse.rs:3]:
```suggestion
/// An iterator adapter that places a separator between all elements.
///
/// This `struct` is created by [`Iterator::intersperse`]. See its
/// documentation for more.
```
Document how was in created.

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [inline library/core/src/iter/adapters/intersperse.rs:11]: I believe `iter` should comes first since it is the most important thing?
```suggestion
    iter: Peekable<I>,
    separator: I::Item,
```

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [inline library/core/src/iter/adapters/intersperse.rs:55]: Wouldn't this result in `sep, item, sep, item` if `next()` was already called once before `fold` on iterator with three `item`?

File: library/core/src/iter/adapters/intersperse.rs
Comment (pickfire) [inline library/core/src/iter/adapters/intersperse.rs:73]: Would this be better?
```suggestion
        let next_is_elem = !self.needs_sep as usize;
        let lo = lo.saturating_sub(next_is_elem).saturating_add(lo);
        let hi = match hi {
            Some(hi) => hi.saturating_sub(next_is_elem).checked_add(hi),
            None => None,
        };
```

File: library/core/tests/iter.rs
Comment (pickfire) [inline library/core/tests/iter.rs:3229]:
```suggestion
    let v: Vec<&str> = xs.iter().cloned().intersperse(", ").collect();
```

File: library/core/tests/iter.rs
Comment (pickfire) [inline library/core/tests/iter.rs:3234]:
```suggestion
    let mut it = ys[..0].iter().copied().intersperse(1);
```
> // copied is the same as .map(|&x| x)

Not sure if this works but probably it does.

File: library/core/tests/iter.rs
Comment (pickfire) [inline library/core/tests/iter.rs:3249]: Should we add a test for one without upper bound?

File: library/core/tests/iter.rs
Comment (pickfire) [inline library/core/tests/iter.rs:3261]: I don't understand what is the difference between these and the above test?

File: library/core/tests/iter.rs
Comment (pickfire) [inline library/core/tests/iter.rs:3255]: Do we need to `clone`?

Comment (pickfire) [inline library/core/tests/iter.rs:3255]: Should we add a test for `fold` and `try_fold` after one item is consumed?

File: library/core/src/iter/traits/iterator.rs
Comment (pickfire) [reply library/core/src/iter/traits/iterator.rs:549]: Haha, how about the second example for `&[u8]`? At least we don't just show `&str`.

File: library/core/src/iter/adapters/intersperse.rs
Comment (LukasKalbertodt) [inline library/core/src/iter/adapters/intersperse.rs:73]: No hard opinion, but this should work as well:

```suggestion
        let hi = hi.and_then(|hi| hi.saturating_sub(next_is_elem as usize).checked_add(hi));
```

File: library/core/src/iter/adapters/intersperse.rs
Comment (LukasKalbertodt) [reply library/core/src/iter/adapters/intersperse.rs:42]: You have removed `try_fold` again, but not the tests. Any reason for that?

File: library/core/src/iter/traits/iterator.rs
Comment (LukasKalbertodt) [reply library/core/src/iter/traits/iterator.rs:549]: Another example where `separator` is yielded more than once would be nice. And preferably with slices instead of strings.

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
It seems like most people are in favor of adding this despite the conflict with `itertools`. I'm also fine with it. But recently it seems like these kinds of conflicts popped up a lot, so, as others have said, a proper lang solution would be really great.

I left a few inline comments, but nothing serious.

What worries me a bit about this API is that it encourages lots of `clone`s and makes it easy to accidentally clone a bunch. Sure, if you pay attention, you probably can avoid expensive clones by only using `Copy` separators. But we all know how "if you pay attention" works out long term in large projects.

However, I don't know how to improve this :/

- Using `Copy` instead of `Clone` as bound seems overly restrictive. Sometimes people might actually want to clone.
- Forcing `I: Iterator<Item = &Separator>` could work but is also kinda awkward? And might not even work as iterators can't return references to `self`.
- Using `Borrow` bounds and yielding `Cow`s (owned for items, borrowed for iterator) makes the API way too complicated.

**Comment by Dylan-DPC** [comment]
@jonas-schievink if you can resolve the conflict, we can get this reviewed quickly :)

**Comment by jonas-schievink** [comment]
I probably won't have time to drive this further and address all the comments, so closing. In case anyone wants to finish this, feel free!

**Comment by Dylan-DPC** [comment]
Thanks :)

**Pull request closed without merging**
